### ISSUE TITLE
Deployment of Trained Model with 3D Velodyne Lidar to real Robot

### ISSUE THREAD

**Issue by nfhe** [reiniscimurs/DRL-robot-navigation#181]
## Description
I've successfully trained a model using the 3D Velodyne lidar within the `DRL-robot-navigation` project. The training process was carried out using the following commands:
```shell
$ cd ~/DRL-robot-navigation/TD3
$ python3 train_velodyne_td3.py
```
To monitor the training progress, I utilized TensorBoard with the command:
```shell
$ cd ~/DRL-robot-navigation/TD3
$ tensorboard --logdir runs
```
Upon completion of the training, I tested the model using:
```shell
$ cd ~/DRL-robot-navigation/TD3
$ python3 test_velodyne_td3.py
```

## Question
I'm currently unsure about how to deploy the trained model (`TD3_velodyne_actor.pth` and `TD3_velodyne_critic.pth`) to a physical robot equipped with a 3D Velodyne lidar. 

### Specific Inquiries
1. Can I directly deploy these trained model files to the physical robot?
2. If direct deployment isn't possible, what steps should I follow for the deployment?
3. Are there any specific modifications or additional configurations required in the code?

## Environment Information
- **Training Environment**: Gazebo simulation environment with the `multi_robot_scenario.launch` file.
- **Model**: Twin Delayed Deep Deterministic Policy Gradient (TD3)
- **Lidar**: 3D Velodyne lidar

## Code Context
The training code primarily involves the following files and their associated functions:

### `train_velodyne_td3.py`
- **Functionality**: Contains the training loop, network initialization, and the implementation of the `TD3` class.

### `test_velodyne_td3.py`
- **Functionality**: Used for testing the trained model.

### `velodyne_env.py`
- **Functionality**: Defines the Gazebo environment, which includes state observation, action execution, and reward calculation.

## Steps to Reproduce (Training and Testing)
1. **Installation**:
    - Install the main dependencies:
        - [ROS Noetic](http://wiki.ros.org/noetic/Installation)
        - [PyTorch](https://pytorch.org/get-started/locally/)
        - [Tensorboard](https://github.com/tensorflow/tensorboard)
    - Clone the repository:
```shell
$ cd ~
$ git clone https://github.com/reiniscimurs/DRL-robot-navigation
```
    - Compile the workspace:
```shell
$ cd ~/DRL-robot-navigation/catkin_ws
$ catkin_make_isolated
```
    - Set up the environment variables and source the workspace:
```shell
$ export ROS_HOSTNAME=localhost
$ export ROS_MASTER_URI=http://localhost:11311
$ export ROS_PORT_SIM=11311
$ export GAZEBO_RESOURCE_PATH=~/DRL-robot-navigation/catkin_ws/src/multi_robot_scenario/launch
$ source ~/.bashrc
$ cd ~/DRL-robot-navigation/catkin_ws
$ source devel_isolated/setup.bash
```
2. **Training**:
```shell
$ cd ~/DRL-robot-navigation/TD3
$ python3 train_velodyne_td3.py
```
3. **Monitoring Training Progress**:
```shell
$ cd ~/DRL-robot-navigation/TD3
$ tensorboard --logdir runs
```
4. **Testing**:
```shell
$ cd ~/DRL-robot-navigation/TD3
$ python3 test_velodyne_td3.py
```

## Expected Behavior
I expect to receive detailed guidance on how to deploy the trained model to the physical robot. This includes clear instructions on whether direct deployment is feasible, and if not, a step-by-step guide for the deployment process. Additionally, I'm looking for information on any necessary code modifications or additional configurations.

## Additional Information
- **OS**: Ubuntu 20.04
- **ROS Version**: ROS Noetic
- **Repository Branch**: Default branch

Any guidance or suggestions would be greatly appreciated!

**Comment by reiniscimurs** [comment]
Hi,

See the answer here: https://github.com/reiniscimurs/DRL-robot-navigation/issues/180

Generally, the test file will load model and get an action in continuous execution. In real life you will not need such functions as reset and reward calculation. So you should create your own environment class with init, velodyne_callback, odom and step methods similarly as in velodyne_env.py file. Set up your subscribers and publishers according to your robot and system and you should be good to go. As long as your input and output space is similar to what is given in the repository, you should not see any major issues.

**Comment by nfhe** [comment]
Dear reiniscimurs,

I sincerely appreciate your previous response. It has been really helpful. 

I have some additional questions regarding LiDAR data. Firstly, I'm curious whether the input from 2D and 3D LiDARs has different impacts on the training process. I've noticed that many people opt for 2D LiDARs, and I'm wondering what factors contribute to this choice in terms of training.

Secondly, in the Gazebo simulation, the LiDAR range is set from -1.57 to 1.57, while in real - world LiDARs, it typically ranges from 0 to 6.28. I'm unsure if an angular transformation of the LiDAR data is necessary in such a case. 

Could you kindly offer your expertise and guidance on these matters?

Thank you very much for your time and assistance.

**Comment by reiniscimurs** [comment]
There is no major difference between 2D and 3D LiDAR data in the training but the 3D lidar allows to capture more complete picture. Lets asuume a situation with an obstacle at 0.2 meter height and your 2D lidar is installed on a plane of 0.4 meters. The robot could trip over and colide with such an obstacle that would cause a different state. The 2D lidar would never the detect it but even the flattened 3D lidar would have this information. However, from the Neural network standpoint nothing changes, since it is still the same 1D input in the state.

The angle does not matter and depends entirely on how you install your lidar. The only thing that matters is that the FOV is 180 degrees in front of the robot. You can adjust the angles where this is captured accordingly.

**Comment by nfhe** [comment]
Dear reiniscimurs,
﻿
Thank you very much for your detailed explanation. It has already provided me with a lot of useful information. However, I still have a few follow - up questions.
﻿
Regarding the 3D LiDAR, since the installation position of 3D LiDAR can vary, and different installation positions result in different point clouds. Does this mean that the installation position during the actual deployment needs to be the same as that during training?
﻿
As for the FOV 180 - degree angle issue you mentioned in the context of 2D and 3D LiDARs, I'm still a bit confused. What I'm specifically concerned about is that the LiDAR range in Gazebo is from - 1.57 to 1.57, while for real - world LiDARs, it's from 0 to 6.28. After clipping the real - world LiDAR data, I get data in the ranges of 0 - 90 and 270 - 359. Do I need to align the order of this real - world LiDAR data with the input order in Gazebo?
﻿
I would greatly appreciate it if you could further clarify these points for me.
﻿
Thanks again for your patience and help.

**Comment by reiniscimurs** [comment]
Your lidar should be installed at a similar height and location. The policy will essentially base it's action on the distance from the obstacle. If the "dangerous" distance is different, it may cause some issues. It depends on the robot size and shape itself but in practice you do not need to be super precise.

However you align your laser does not matter, what you need to have is sequential 180 degrees from left to right in front of the robot. In this case a combination of 270-360 and then 0-90 degree range should work. Also just using 0 to 180 degrees work if you place your lidar sideways. We need a continuous sequence of points from left to right for 180 degrees, just like it is in the simulation.

**Comment by nfhe** [comment]
Dear reiniscimurs,

I truly appreciate your patient and detailed responses. They have been extremely helpful to me. I will now follow your advice and conduct tests on both 3D and 2D LiDARs respectively. I'm closing this issue. Thank you once again for all your assistance.
